File: ch7/java-client/src/main/java/com/github/packtpub/cntkquickstart/FlowerClassifier.java
package com.github.packtpub.cntkquickstart;

import com.microsoft.CNTK.*;

public class FlowerClassifier {
    private final DeviceDescriptor deviceDescriptor;
    private final Function modelFunction;

    private FlowerClassifier(Function modelFunction, DeviceDescriptor deviceDescriptor) {
        this.deviceDescriptor = deviceDescriptor;
        this.modelFunction = modelFunction;
    }

    public float[] predict(float petalWidth, float petalLength, float sepalWidth, float sepalLength) {
        Variable features = modelFunction.getInputs().get(0);
        Variable output = modelFunction.getOutputs().get(0);

        FloatVectorVector batch = new FloatVectorVector();

        batch.add(createFloatVector(sepalLength, sepalWidth, petalLength, petalWidth));

        UnorderedMapVariableValuePtr inputMapping = new UnorderedMapVariableValuePtr();
        UnorderedMapVariableValuePtr outputMapping = new UnorderedMapVariableValuePtr();

        inputMapping.add(features, Value.createDenseFloat(features.getShape(), batch, deviceDescriptor));
        outputMapping.add(output, null);

        modelFunction.evaluate(inputMapping, outputMapping, deviceDescriptor);

        FloatVectorVector outputBuffer = new FloatVectorVector();
        outputMapping.getitem(output).copyVariableValueToFloat(output, outputBuffer);

        FloatVector predictedOutput = outputBuffer.get(0);

        return getVectorValues(predictedOutput);
    }

    private static float[] getVectorValues(FloatVector predictedOutput) {
        float[] outputVector = new float[(int) predictedOutput.size()];

        for (int i = 0; i < predictedOutput.size(); i++) {
            outputVector[i] = predictedOutput.get(i);
        }

        return outputVector;
    }

    private static FloatVector createFloatVector(float... inputValues) {
        FloatVector vector = new FloatVector();

        for (float inputValue : inputValues) {
            vector.add(inputValue);
        }

        return vector;
    }

    public static FlowerClassifier create() {
        DeviceDescriptor cpuDevice = DeviceDescriptor.getCPUDevice();
        Function modelFunction = Function.load("model.onnx", cpuDevice, ModelFormat.ONNX);

        return new FlowerClassifier(modelFunction, cpuDevice);
    }
}

File: ch7/csharp-client/FlowerClassifier.cs
using System.Collections.Generic;
using CNTK;

namespace PredictSpecies
{
    public class FlowerClassifier
    {
        private readonly DeviceDescriptor _deviceDescriptor;
        private readonly Function _modelFunction;

        private FlowerClassifier(Function modelFunction, DeviceDescriptor deviceDescriptor)
        {
            _deviceDescriptor = deviceDescriptor;
            _modelFunction = modelFunction;
        }

        public IList<float> Predict(float petalWidth, float petalLength, float sepalWidth, float sepalLength)
        {
            var features = _modelFunction.Inputs[0];
            var output = _modelFunction.Outputs[0];

            var inputMapping = new Dictionary<Variable, Value>();
            var outputMapping = new Dictionary<Variable, Value>();

            var batch = Value.CreateBatch(
                features.Shape,
                new float[] { sepalLength, sepalWidth, petalLength, petalWidth },
                _deviceDescriptor);

            inputMapping.Add(features, batch);
            outputMapping.Add(output, null);

            _modelFunction.Evaluate(inputMapping, outputMapping, _deviceDescriptor);

            var outputValues = outputMapping[output].GetDenseData<float>(output);

            return outputValues[0];
        }

        public static FlowerClassifier Create()
        {
            var deviceDescriptor = DeviceDescriptor.CPUDevice;
            var function = Function.Load("model.onnx", deviceDescriptor);

            return new FlowerClassifier(function, deviceDescriptor);
        }

        private static FloatVector CreateFloatVector(params float[] inputValues)
        {
            FloatVector vector = new FloatVector();

            foreach (var inputValue in inputValues)
            {
                vector.Add(inputValue);
            }

            return vector;
        }
    }
}